### PR TITLE
mgr/dashboard: reduce method (cyclomatic) complexity 

### PR DESCRIPTION
**Pull request by VallariAg**
Fixes: https://tracker.ceph.com/issues/41221

I resolved all complexity warnings under `flake8 --max-complexity 10 src/pybind/mgr/dashboard/controllers/`.

Help/guidance needed: I am not sure how to test my changes. The `tox` tests I ran just checked it for linting problems and I fixed them. The other tests I was trying to run were [these](https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#run-cephadm-e2e-tests-sh) dashboard's e2e tests. But I'm getting this error 
```
2022-02-14T19:26:27Z Eyes: [LOG    ] {} (): starting plugin server
2022-02-14T19:26:27Z Eyes: [LOG    ] {} Server.<init>(): plugin server running at port: 46455
Cypress could not verify that this server is running:

  > http://localhost:4200/

We are verifying this server because it has been configured as your `baseUrl`.

Cypress automatically waits until your server is accessible before running tests.

We will try connecting to it 3 more times...
We will try connecting to it 2 more times...
We will try connecting to it 1 more time...

Cypress failed to verify that your server is running.

Please start this server and then run Cypress again.

```
I do have the cluster running with vstart as written in README and dashboard frontend is running fine at https://127.0.0.1:41354

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by nizamial09** [comment]
> The `tox` tests I ran just checked it for linting problems and I fixed them.

were you running[ these ](https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#unit-tests-based-on-tox)? 

> The other tests I was trying to run were [these](https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#run-cephadm-e2e-tests-sh) dashboard's e2e tests.

The link you pointed is to run the dashboard cephadm checks and I think probably that's not needed over here. What you can do is to try to run this [dashboard e2e check ](https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#run-frontend-e2e-tests-sh)instead.

If you are interested, we use a developer environment called[ ceph-dev ](https://github.com/rhcs-dashboard/ceph-dev) which is much more easier to setup and also in the [README ](https://github.com/rhcs-dashboard/ceph-dev/blob/master/README.md)file you'll find all sorts of commands for running the tests easily.

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by VallariAg** [comment]
> were you running[ these ](https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#unit-tests-based-on-tox)?

yes! I was referring to [this](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/#running-tox) similar page.

> If you are interested, we use a developer environment called[ ceph-dev ](https://github.com/rhcs-dashboard/ceph-dev)which is much more easier to setup and also in the [README ](https://github.com/rhcs-dashboard/ceph-dev/blob/master/README.md)file you'll find all sorts of commands for running the tests easily.

This looks perfect, I'll set this up, thank you!

File: src/pybind/mgr/dashboard/controllers/_base_controller.py
Comment (epuertat) [inline src/pybind/mgr/dashboard/controllers/_base_controller.py:272]: We [already have a method](https://github.com/VallariAg/ceph/blob/343b580be12d760cbd258da6f02689552a68180f/src/pybind/mgr/dashboard/controllers/_version.py#L50-L55) very similar to this one.

**Comment by epuertat** [comment]
jenkins retest this please

**Review by epuertat — COMMENTED** [review]
Nice clean-up! Thanks a lot @VallariAg. Let's wait for the CI tests to pass... Perhaps they were failing due to something over here?

File: src/pybind/mgr/dashboard/controllers/_base_controller.py
Comment (VallariAg) [reply src/pybind/mgr/dashboard/controllers/_base_controller.py:272]: Oh, awesome, I'll import that!

**Comment by VallariAg** [comment]
hey @epuertat, thank you for reviewing this!
I've been testing the code with `ceph-dev` frontend e2e tests scripts, like @nizamial09 suggested. I found a few more errors already. 
I'm trying to get all testcases green there locally. Will commit and push my progress soon!

**Review by avanthakkar — CHANGES_REQUESTED** [review]

**Comment by avanthakkar** [comment]
@VallariAg The rgw client tests needs to be adopted and for fixing the dashboard lint errors you can run `tox -e fix`  under `src/pybind/mgr/dashboard`

```RgwClientTest.test_set_bucket_locking_error __________________

self = <dashboard.tests.test_rgw_client.RgwClientTest testMethod=test_set_bucket_locking_error>

    def test_set_bucket_locking_error(self):
        instance = RgwClient.admin_instance()
        test_params = [
            ('COMPLIANCE', 'null', None, 'must be a positive integer'),
            ('COMPLIANCE', None, 'null', 'must be a positive integer'),
            ('COMPLIANCE', -1, None, 'must be a positive integer'),
            ('COMPLIANCE', None, -1, 'must be a positive integer'),
            ('COMPLIANCE', 1, 1, 'You can\'t specify both at the same time'),
            ('COMPLIANCE', None, None, 'You must specify at least one'),
            ('COMPLIANCE', 0, 0, 'You must specify at least one'),
            (None, 1, 0, 'must be either COMPLIANCE or GOVERNANCE'),
            ('', 1, 0, 'must be either COMPLIANCE or GOVERNANCE'),
            ('FAKE_MODE', 1, 0, 'must be either COMPLIANCE or GOVERNANCE')
        ]
        for params in test_params:
            mode, days, years, error_msg = params
            with self.assertRaises(DashboardException) as cm:
>               instance.set_bucket_locking(
                    bucket_name='test',
                    mode=mode,
                    retention_period_days=days,
                    retention_period_years=years
                )

tests/test_rgw_client.py:248: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
rest_client.py:529: in func_wrapper
    return func(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @RestClient.api_put('/{bucket_name}?object-lock')
    def set_bucket_locking(self,
                           bucket_name: str,
                           mode: str,
                           retention_period_days: Optional[Union[int, str]] = None,
                           retention_period_years: Optional[Union[int, str]] = None,
                           request: Optional[object] = None) -> None:
        """
        Places the locking configuration on the specified bucket. The
        locking configuration will be applied by default to every new
        object placed in the specified bucket.
        :param bucket_name: The name of the bucket.
        :type bucket_name: str
        :param mode: The lock mode, e.g. `COMPLIANCE` or `GOVERNANCE`.
        :type mode: str
        :param retention_period_days:
        :type retention_period_days: int
        :param retention_period_years:
        :type retention_period_years: int
        :rtype: None
        """
        # pylint: disable=unused-argument
    
>       retention_period_days = int(retention_period_days) if retention_period_days else 0
E       ValueError: invalid literal for int() with base 10: 'null'```

**Comment by avanthakkar** [comment]
jenkins test windows

File: src/pybind/mgr/dashboard/services/tcmu_service.py
Comment (epuertat) [reply src/pybind/mgr/dashboard/services/tcmu_service.py:20]: Interesting... pylint (or some static checked) should have reported this 🤔

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/services/tcmu_service.py:20]: > Interesting... pylint (or some static checked) should have reported this thinking

Yes, mypy did!

File: src/pybind/mgr/dashboard/controllers/_base_controller.py
Comment (pereman2) [inline src/pybind/mgr/dashboard/controllers/_base_controller.py:242]: Is this reducing the cyclomatic complexity? Theoretically, this is the exact same in cyclomatic complexity terms, but using ternary operators might trick mypy.

Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/_base_controller.py:242]: Yup, I had the same feeling when I saw this. Perhaps it's reducing the number of lines (but AFAICR McCabe's doesn't account for the LoC), and hence Pylint score is higher.

Comment (VallariAg) [reply src/pybind/mgr/dashboard/controllers/_base_controller.py:242]: Hmm, I see what you mean! I thought it did improve the complexity because it updated value of one variable which could be similar to making a new function like this:
```python3
def unquote_value(value):
    return unquote(value) if isinstance(value, str) else value

# then calling this like: 
kwargs[key] = unquote_value(value)
```
But yes, theoretically, using ternary operators should not change cyclomatic complexity because the if-else conditions still adds a new path in the control flow graph for this method. I guess flake8 has been tricked somehow and it's not an improved complexity!
I'll look for more meaningful ways to improve the score, thanks for pointing that out.

**Comment by epuertat** [comment]
jenkins test make check

**Comment by epuertat** [comment]
jenkins test windows

File: src/pybind/mgr/dashboard/controllers/_base_controller.py
Comment (pereman2) [reply src/pybind/mgr/dashboard/controllers/_base_controller.py:242]: @VallariAg Sorry I didn't notice you left a comment, are there any updates on this? Anyways, when make check passes I'm ok approving

**Comment by epuertat** [comment]
jenkins test make check

**Comment by VallariAg** [comment]
make check passed without any changes? 
@epuertat the dashboard tests are all passing locally for me

**Comment by VallariAg** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
> how did make check pass without any changes? @epuertat the dashboard tests are all passing as well locally for me

Unfortunately tests (especially e2e) aren't 100% stable, so if you see some failure unrelated to your changes, feel free to retrigger that.

**Comment by epuertat** [comment]
In this specific case, what's failing in the e2e is a visual comparison between baseline screenshots and screenshots taken from this PR, but the change is unrelated to this, so marked those as OK, and hence this should now pass.

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by VallariAg** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
jenkins test dashboard

**Review by epuertat — APPROVED** [review]
LGTM! Thanks a lot @VallariAg !

**Review by pereman2 — APPROVED** [review]

**Review by avanthakkar — APPROVED** [review]

**Comment by epuertat** [comment]
Merged. Congrats, @VallariAg ! 🥳

**Comment by ljflores** [comment]
@epuertat this might be causing a test failure in the rados suite: http://pulpito.front.sepia.ceph.com/yuriw-2022-04-22_13:56:48-rados-wip-yuri2-testing-2022-04-22-0500-distro-default-smithi/6800373/

ceph-mgr.x.log.gz
```
2022-04-22T17:18:14.253+0000 7fcf90831700  0 [dashboard DEBUG auth] checking authorization...
2022-04-22T17:18:14.253+0000 7fcf90831700  0 [dashboard ERROR exception] Internal Server Error
Traceback (most recent call last):
  File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 238, in get_client_version
    cherrypy.request.headers['Accept'])
  File "/usr/share/ceph/mgr/dashboard/controllers/_version.py", line 41, in from_mime_type
    return cls.from_string(cls.__MIME_TYPE_REGEX.match(mime_type).group(1))
AttributeError: 'NoneType' object has no attribute 'group'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/share/ceph/mgr/dashboard/services/exception.py", line 47, in dashboard_exception_handler
    return handler(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/cherrypy/_cpdispatch.py", line 54, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 260, in inner
    client_version = BaseController.get_client_version()
  File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 241, in get_client_version
    415, "Unable to find version in request header")
cherrypy._cperror.HTTPError: (415, 'Unable to find version in request header')
```

I'm looking into this to see if it reproduces. Do you know if this PR went through a final test before it was merged?

**Comment by epuertat** [comment]
> @epuertat this might be causing a test failure in the rados suite: [http://pulpito.front.sepia.ceph.com/yuriw-2022-04-22_13:56:48-rados-wip-yuri2-testing-2022-04-22-0500-distro-default-smithi/6800373](http://pulpito.front.sepia.ceph.com/yuriw-2022-04-22_13:56:48-rados-wip-yuri2-testing-2022-04-22-0500-distro-default-smithi/6800373/)
> 
> ceph-mgr.x.log.gz
> 
> ```
> 2022-04-22T17:18:14.253+0000 7fcf90831700  0 [dashboard DEBUG auth] checking authorization...
> 2022-04-22T17:18:14.253+0000 7fcf90831700  0 [dashboard ERROR exception] Internal Server Error
> Traceback (most recent call last):
>   File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 238, in get_client_version
>     cherrypy.request.headers['Accept'])
>   File "/usr/share/ceph/mgr/dashboard/controllers/_version.py", line 41, in from_mime_type
>     return cls.from_string(cls.__MIME_TYPE_REGEX.match(mime_type).group(1))
> AttributeError: 'NoneType' object has no attribute 'group'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/share/ceph/mgr/dashboard/services/exception.py", line 47, in dashboard_exception_handler
>     return handler(*args, **kwargs)
>   File "/usr/lib/python3.6/site-packages/cherrypy/_cpdispatch.py", line 54, in __call__
>     return self.callable(*self.args, **self.kwargs)
>   File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 260, in inner
>     client_version = BaseController.get_client_version()
>   File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 241, in get_client_version
>     415, "Unable to find version in request header")
> cherrypy._cperror.HTTPError: (415, 'Unable to find version in request header')
> ```
> 
> I'm looking into this to see if it reproduces. Do you know if this PR went through a final test before it was merged?

Hi @ljflores, this PR passed all the Dashboard core testing (not the teuthology one though, but the Ceph API test should be more thorough when catching REST API version issues). Anyway, in any Ceph API test run you'll see a trace like that, since as part of the testing, version mismatches are triggered. Given those are expected failures, we should replace the tracebacks with one-line log messages (and leave the tracebacks for the real unexpected ones).

@avanthakkar if needed, can you please help @ljflores debug this issue? Thanks!

**Comment by ljflores** [comment]
> Hi @ljflores, this PR passed all the Dashboard core testing (not the teuthology one though, but the Ceph API test should be more thorough when catching REST API version issues). Anyway, in any Ceph API test run you'll see a trace like that, since as part of the testing, version mismatches are triggered. Given those are expected failures, we should replace the tracebacks with one-line log messages (and leave the tracebacks for the real unexpected ones).
> 
> @avanthakkar if needed, can you please help @ljflores debug this issue? Thanks!

Thanks for the context @epuertat! I'll create a Tracker so the discussion can continue there.

**Pull request merged**
